esp-hal/src/dma/mod.rs

```diff
@@ -1837,7 +1837,12 @@ pub struct Preparation {
 
 /// [DmaTxBuffer] is a DMA descriptor + memory combo that can be used for
 /// transmitting data from a DMA channel to a peripheral's FIFO.
-pub trait DmaTxBuffer {
+///
+/// # Safety
+///
+/// The implementing type must keep all its descriptors and the buffers they
+/// point to valid between each call to [Self::prepare].
```

Comment:
`valid between each call to [Self::prepare].`

This wording suggests "all the time" because it is not defined when the buffer's resources are allowed to be freed, or otherwise invalidated.

Reply:
Ah fair, it can be freed when the object is dropped. I'll update

Reply:
So does that mean the user can call `prepare()` and immediately drop the buffer?

Reply:
Yes that's allowed. So long as the DMA is holding the implementation of the trait, all the descriptors and buffers they point to should remain valid (not necessarily stable).

Reply:
Mmm, okay, it's still somewhat unclear what "valid" means in this context, though. There are two things I want to try, both of which might bump up against the wording:

- What can I do with a buffer if I add a `swap_tx/rx/_buffer` API to SpiDmaBus to allow preparing a buffer while a transfer is in progress?
- How can I implement a dynamically allocated buffer? It will likely reallocate the buffer, so the old and new `prepare` calls will return something else, and the old buffer will likely be freed. At which point am I, as the implementor, allowed to invalidate the pointer to the buffer?

There are "technically fine because the DMA owns the buffer" points here, but since this trait is now unsafe, I need the wording to be very precise.

Reply:
> it's still somewhat unclear what "valid" means in this context, though

I agree "valid" is vague. I was being lazy there.

> * What can I do with a buffer if I add a `swap_tx/rx/_buffer` API to SpiDmaBus to allow preparing a buffer while a transfer is in progress?

I'm not sure I understand this, so I'll answer a slightly different question, "How can I prepare a buffer while a transfer is in progress". `SpiDmaBus` would need to change to accept two (of each type TX/RX) buffers, one to fill and one to be "in progress". (I will write up an example now that you ask, it would look something like https://github.com/Dominaezzz/esp-hal/blob/616b2dd74618421b169970c1665f126d97b64a09/examples/src/bin/spi_loopback_pipeline.rs#L51-L76)

> * How can I implement a dynamically allocated buffer? It will likely reallocate the buffer, so the old and new `prepare` calls will return something else, and the old buffer will likely be freed. At which point am I, as the implementor, allowed to invalidate the pointer to the buffer?

This is a good question because it gets to the heart of what should eventually be in the safety doc.
You're allowed to re-allocate the buffer, even whilst the DMA is still using the buffer but there are rules. (I do this in one of my projects)

1. The pointer to the descriptor returned by `prepare` (I call this the "head" descriptor) must point to valid (intialized but not destroyed) memory  until the next call to `prepare` or `drop`.
2. The descriptors that the head descriptor point to are allowed to change at any time.
3. The buffer pointed to by a descriptor in the linked list following the head descriptor (I shall refer to this linked list as the "descriptor chain") must be valid.
4. If a descriptor is removed from the descriptor chain, it and the descriptors it points to must still follow the rules above until you can prove the DMA can no longer reach them, i.e. using the owner bit, EOF bit, stopping/restarting the transfer or some other means.

Unfortunately the above requires you to really know what you're doing, and grok the "Linked List Reading Process" of the TRMs. Eventually there should be docs here to re-iterate that stuff I guess.
 
> This wording suggests "all the time" because it is not defined when the buffer's resources are allowed to be freed, or otherwise invalidated.

This is the main reason I didn't write the stuff I just commented in the safety doc. At the moment you can't (easily) access a buffer whilst it's in use by the DMA/peripheral driver since it takes full ownership.
Though I have a vague plan to allow some kind of access, since circular transfers kind of need it.

```Rust
pub unsafe trait DmaTxBuffer {
    type View<'a>;

    fn prepare(&mut self) -> Preparation;

    fn make_view(&mut self) -> View<'a>;
}
```

The plan is to add an associated type `View` to the buffer trait, and this `View` type will expose a limited set of operations that are allowed to be performed whilst the DMA is reading the descriptor linked list returned by `prepare`.

Reply:
Let's relax the requirement to "The implementing type must keep all its descriptors and the buffers they point to valid while the buffer is being transferred", and let's deal with the slightly insane case of "modifying a buffer in-flight" later. Much, much later :D Unfortunately we don't exactly know when the buffer is no longer being transferred in this API, so maybe a callback `unsafe fn transfer_complete` would be nice in this case, that the drivers could call.

Reply:
SGTM, will update.

> so maybe a callback `unsafe fn transfer_complete` would be nice in this case, that the drivers could call.

I've been vaguely considering using type state to solve this problem but we can dig into this later.